core/src/main/java/org/geysermc/geyser/extension/command/GeyserExtensionCommand.java

```diff
@@ -25,19 +25,206 @@
+        }
+
+        @Override
+        public Builder<T> permission(@NonNull String permission) {
+            this.permission = Objects.requireNonNull(permission, "command permission");
+            return this;
+        }
+
+        @Override
+        public Builder<T> permission(@NonNull String permission, @NonNull TriState defaultValue) {
+            this.permission = Objects.requireNonNull(permission, "command permission");
+            this.permissionDefault = Objects.requireNonNull(defaultValue, "command permission defaultValue");
+            return this;
+        }
+
+        @Override
+        public Builder<T> aliases(@NonNull List<String> aliases) {
+            this.aliases = aliases;
+            return this;
+        }
+
+        @Override
+        public Builder<T> suggestedOpOnly(boolean suggestedOpOnly) {
+            this.suggestedOpOnly = suggestedOpOnly;
+            if (suggestedOpOnly) {
+                // the most amount of legacy/deprecated behaviour I'm willing to support
+                this.permissionDefault = TriState.NOT_SET;
+            }
+            return this;
+        }
+
+        @Override
+        public Builder<T> executableOnConsole(boolean executableOnConsole) {
+            this.playerOnly = !executableOnConsole;
+            return this;
+        }
+
+        @Override
+        public Command.Builder<T> playerOnly(boolean playerOnly) {
+            this.playerOnly = playerOnly;
+            return this;
+        }
+
+        @Override
+        public Builder<T> bedrockOnly(boolean bedrockOnly) {
+            this.bedrockOnly = bedrockOnly;
+            return this;
+        }
+
+        @Override
+        public Builder<T> executor(@NonNull CommandExecutor<T> executor) {
+            this.executor = executor;
+            return this;
+        }
+
+        @NonNull
+        @Override
+        public GeyserExtensionCommand build() {
+            // These are captured in the anonymous lambda below and shouldn't change even if the builder does
+            final Class<? extends T> sourceType = this.sourceType;
+            final boolean suggestedOpOnly = this.suggestedOpOnly;
+            final CommandExecutor<T> executor = this.executor;
+
+            if (name == null) {
+                throw new IllegalArgumentException("name was not provided for a command in extension " + extension.name());
+            }
+            if (sourceType == null) {
+                throw new IllegalArgumentException("Source type was not defined for command " + name + " in extension " + extension.name());
+            }
+            if (executor == null) {
+                throw new IllegalArgumentException("Command executor was not defined for command " + name + " in extension " + extension.name());
+            }
+
+            // if the source type is a GeyserConnection then it is inherently bedrockOnly
+            final boolean bedrockOnly = this.bedrockOnly || GeyserConnection.class.isAssignableFrom(sourceType);
```

Comment:
can we also warn about this?

Reply:
warn if they set the source type as a GeyserConnection but `bedrockOnly` wasn't specified (or specified as false, which results in the same value of `this.bedrockOnly`) ?

Reply:
yeah; just to make it obvious

Reply:
i can agree somewhat. however, requiring a `GeyserConnection` is already quite obvious that it's only going to be bedrock players. requiring them to write `bedrockOnly(true)` just adds verbosity and congests logs

Reply:
It shouldn't really spam anything if the api is used correctly?
Also, it would only be sent once when the command is built :p

While it's fairly obvious to us that a GeyserConnection is effectively making the command bedrock only, setting that variable is imo more clearer. It should, at the very least, make it easier to spot the issue for people creating extension commands

Reply:
> It shouldn't really spam anything if the api is used correctly?
it's not really being used incorrectly if they just don't specify `bedrockOnly`. if they say `bedrockOnly(false)`, then sure, but we'd have to make the field `Boolean` and initially null to check. which i guess i'm okay with but it seems kinda extra.

> it would only be sent once when the command is built
i suppose there doesn't have to be a stacktrace because we know the extension in question